slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -329,6 +330,111 @@ void SLAMServiceImpl::SetUpMapBuilder() {
 }
 
 std::string SLAMServiceImpl::GetLatestJpegMapString(bool add_pose_marker) {
+    try {
+        cartographer::io::PaintSubmapSlicesResult painted_slices =
+            GetLatestPaintedMapSlices();
+    } catch (std::exception &e) {
+        // jpeg will be empty which captures the same error
+        return "";
+    }
+    cartographer::io::PaintSubmapSlicesResult painted_slices =
+        GetLatestPaintedMapSlices();
+    if (add_pose_marker) {
+        PaintMarker(painted_slices);
+    }
+
+    auto image = io::Image(std::move(painted_slices.surface));
+    return image.WriteJpegToString(50);
+}
+
+bool SLAMServiceImpl::GetLatestPointCloudMapString(std::string &pointcloud) {
+    try {
+        cartographer::io::PaintSubmapSlicesResult painted_slices =
+            GetLatestPaintedMapSlices();
+    } catch (std::exception &e) {
+        // pointcloud will be empty which captures the error
+        return false;
+    }
+    cartographer::io::PaintSubmapSlicesResult painted_slices =
+        GetLatestPaintedMapSlices();
+    auto painted_surface = painted_slices.surface.get();
+    int width = cairo_image_surface_get_width(painted_surface);
+    int height = cairo_image_surface_get_height(painted_surface);
+    auto data = cairo_image_surface_get_data(painted_surface);
+
+    // total number of bytes in image (4 bytes per pixel)
+    int size_data = width * height * 4;
+
+    std::vector<unsigned char> data_vect(data, data + size_data);
+
+    int num_points = 0;
+
+    // scaling factor to ensure number of pixels is within required size
+    // do not set to 0
+    float scaleFactor = .5;
+
+    // Reduce resolution based off number of pixels. Output is pixels to skip
+    // Ideally should be between 5 and 15, but depends on the resolution of the
+    // original image
+    int scale = size_data / maximumBitLimit * scaleFactor;
+
+    std::string data_buffer;
+
+    // Loop to filter unwanted data and reduce resolution. Increments multiplied
+    // by 4 to represent 4 bytes per pixel
+    for (int i = 0; i < size_data; i += scale * 4) {
+        // skip pixels that are not in our map(black/past walls)
+        // this check represents [102,102,102]
+        if (((int)data_vect[i + 0] == defaultRGBValue) &&
+            ((int)data_vect[i + 1] == defaultRGBValue) &&
+            ((int)data_vect[i + 2] == defaultRGBValue))
+            continue;
+
+        // Determine probability based off color pixel
+        int prob = ViamColorToProbability((int)data_vect[i + 2]);
+        if (prob == 0) continue;
+
+        num_points++;
+
+        int pixel_index = i / 4;
+        int pixel_x = pixel_index % width;
+        int pixel_y = pixel_index / width;
+
+        // based on PaintSubmapSlices() and DrawPoseOnSurface()
+        float x_pos = (pixel_x - painted_slices.origin.x()) * kPixelSize;
```

Comment:
Could this be commented better / wrapped in a commented function? I don't understand what this is doing, nor do I understand what the units of kPixelSize are.